### PR TITLE
fix miner message filter nonce checking

### PR DESCRIPTION
**Pull request by whyrusleeping**
When mining on top of a tipset that contains a message with nonce X, miners will currently not be able to mine a block with nonce X+1 because they are checking the nonce of the parent state, not the state of the tipset they are mining on.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
The more right way to do this might be to just evaluate the state of the tipset we're mining on and then simply check the nonce... Its a little tricky though.

**Comment by whyrusleeping** [comment]
Actually, looking at how StateGetActor works, it should be doing the right thing. We pass it the tipset we're mining on top of, it calls `stmgr.TipSetState` on that tipset, which evaluates its messages, and returns the resultant state tree, which we then pull the actor from to source the nonce...

That seems correct to me. What bug am i actually fixing here?

**Comment by whyrusleeping** [comment]
To clarify, i'm trying to fix the fact that i see a whole lot of this:
```
2019-12-02T22:08:18.188-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (134 > 133) bafy2bzaced7x7az63ss4giowqywg43revbhwjf4pdl2m4itzvug526f7bk5e2
2019-12-02T22:08:18.189-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (254 > 253) bafy2bzacecquzo4qixfkoecndmnu5vmpmv65qelausjppk6l4za2aqolawkyq
2019-12-02T22:08:18.190-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (236 > 190) bafy2bzacechiyqtvak4bxul4xsaw4qu5n4cvgik7dgozbffiyvjzeaesruuxo
2019-12-02T22:08:18.191-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (243 > 190) bafy2bzacednll3nwv33y45jg7zkan3qhc2kvabygvehzqmb5kkudjytma7if4
2019-12-02T22:08:18.191-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (262 > 190) bafy2bzaceb7tf3bp5fz754tvsvqzgb6kb75ebdibjd5zelbxlg5joqtpphzh6
2019-12-02T22:08:18.192-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (265 > 190) bafy2bzacedhsokrrawt73vzwuv44xjsygttwn2bstxtux2zgm3unsb72ttk5y
2019-12-02T22:08:18.193-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (266 > 190) bafy2bzaceaadnyiw4ajannmdhcmqv33jrk3vgalmrwusfk6f6as4clxubiara
2019-12-02T22:08:18.193-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (267 > 190) bafy2bzaceag32i3p4fgolsoiysgx32eynlnyhzmmgcv7l4t4yso55phxkj2cu
2019-12-02T22:08:18.194-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (268 > 190) bafy2bzacebkd526fjmbh6rncptp7w6p3hykzrq3omgcc7s4fuojkwdj2h6a46
2019-12-02T22:08:18.195-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (269 > 190) bafy2bzacec333ypt24d4ytbkwk3tvupwdtnvvtlir7hheuydngzs4aoha2pje
2019-12-02T22:08:18.196-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (270 > 190) bafy2bzacedp4vcxnnlvonmtd5cj4zckdkv2sb4pqrxhhvecyywcqo34volrvs
2019-12-02T22:08:18.196-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (271 > 190) bafy2bzacecykpu5uunh3r4mvxdxnqra4iasc5vduoy3luhk53rqshh42dr7be
2019-12-02T22:08:18.197-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (272 > 190) bafy2bzacec5kzrtp2xpomgnt4me7pxntvppgw7echcoacl4byppwro4bv2p52
2019-12-02T22:08:18.198-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (273 > 190) bafy2bzacea24lv4t45syzpkkp7amdr5py2s6oxhjjb6wd6fvqdbjrkanjyeuk
2019-12-02T22:08:18.198-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (274 > 190) bafy2bzacebqrldoluok4dpjojaqwzjsbode7h5uzudvqosmluezvglc3upbwm
2019-12-02T22:08:18.199-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (275 > 190) bafy2bzacedkab66vnq7lifcys5tdiotueb4qam7hasjjmrhilpnnzqwvvweeg
2019-12-02T22:08:18.200-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (276 > 190) bafy2bzaceayjm4n7dpgif3pwkl7mafa5mjkbw4so5dqzdzmk4ydmym6xxyfry
2019-12-02T22:08:18.200-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (277 > 190) bafy2bzaceamrtxiqnflecdq2pgfjulvl4ymhsvd3kdzy4zhybqxekjtd6pvbs
2019-12-02T22:08:18.201-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (278 > 190) bafy2bzacecbhiqaiaqaxjl4mgtcpy3u5oho5ptudsjrjwbull22uq2ebr64jm
2019-12-02T22:08:18.202-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (279 > 190) bafy2bzacebjsw3yqa5gor6fuxj7s4vemqlwmrvdfplu5672xawijhgjibv7h6
2019-12-02T22:08:18.202-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (280 > 190) bafy2bzacedblhcg3fk27elxibki5z22vsggmujn2g5jyxevduwops4juu5lzo
2019-12-02T22:08:18.203-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (281 > 190) bafy2bzacedew2jd7b3oeklseqed26ujqzbhanxcehszdngwkfs7skqpvkafwy
2019-12-02T22:08:18.204-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (282 > 190) bafy2bzacecgi77b632dfhzshfyayzyzlhadkddzxynw4wsii22cetmrld4pl2
2019-12-02T22:08:18.204-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (283 > 190) bafy2bzaceaaautyvcqkxnwaxiigehlohrnjdxu36f6mrjni4aqv7ih5lwjz5u
2019-12-02T22:08:18.205-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (284 > 190) bafy2bzacedg63gehfvde44yc4f4e32jblm7esimn6lcxmbt4i53wiuzamnqs2
2019-12-02T22:08:18.206-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (285 > 190) bafy2bzaceawwzutw2ugqwwevzyuef47cyb4pivtbvp2t2d7lo7a466st2dzhg
2019-12-02T22:08:18.207-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (286 > 190) bafy2bzacedd5tbiuxdxgoblnbktitu4467caiculsl3l3fjvmgjrs4ck2uf2k
2019-12-02T22:08:18.207-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (287 > 190) bafy2bzaced6iarsj3ocqnh5hb7k46upmrnbodyniyxcusmpjli55p7eq4t4ci
2019-12-02T22:08:18.208-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (288 > 190) bafy2bzaceddeebs7mztvfqdzm7jumaozftq664jjirmmmc5rn62mps3nlong4
2019-12-02T22:08:18.209-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (289 > 190) bafy2bzacebvlbmcw6hujjwedeyejgsfqeqmvdzf3jtn44fm5i4isogoe5sibw
2019-12-02T22:08:18.209-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (290 > 190) bafy2bzaceagshaalgypcnjx6oh3mv6brhir4ddoxorhzzwm3zop7kpx75e7hu
2019-12-02T22:08:18.210-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (291 > 190) bafy2bzacebprymftpuppd3cy36itny4tehonzd4iw7z3z2oos7paxc3p6ubqo
2019-12-02T22:08:18.211-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (292 > 190) bafy2bzacebkhibd5edx4zfxvrnrs7iinhjvvdztk77alih2taykfqoitxafiw
2019-12-02T22:08:18.211-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (293 > 190) bafy2bzaceb3tbhwseplrypb5fkdthnshzj6lkkxajh4vg346h4z3z256sodsq
2019-12-02T22:08:18.212-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (294 > 190) bafy2bzaceazmw2dtlnujob3qehhzroy5ldufn4h5mxdvvtwzferr34r5a5b5o
2019-12-02T22:08:18.213-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (295 > 190) bafy2bzacec67xy7fs3uzref6zdhjfjtpnjbmwczzsntt4nmidnthlb4evy5ea
2019-12-02T22:08:18.214-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (296 > 190) bafy2bzacea7luly36lldwewfkf6f3no7o7bjr6qep5g77ib5chy3oapu3aico
2019-12-02T22:08:18.214-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (297 > 190) bafy2bzaceckzjjuon27xndlahcdnc4l7z4nol4nsij2pvpw36jrr3olliv6ly
2019-12-02T22:08:18.215-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (298 > 190) bafy2bzacebsrxun3msl33sednqfs6crt6cg7yptlrn3w6zy6kcbqudpkdhu3m
2019-12-02T22:08:18.216-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (299 > 190) bafy2bzacecdhujtnl45xxlvhuznl4cqn7gnyanp3p7jgmuda6kusze36bcg3y
2019-12-02T22:08:18.216-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (300 > 190) bafy2bzaceacfie24bhwlox7jebeoxmi3aijapdoub4m4443a5k6z3t5gdqlxm
2019-12-02T22:08:18.217-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (301 > 190) bafy2bzacedio6nfimq3hsgfsb3wmial6whdceppqzhftj6z2wed6x5sh2zpci
2019-12-02T22:08:18.218-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (302 > 190) bafy2bzacedhjjaclhymsvmqekvqczgo7elrqo6pafyrhq6djp6nmzisk4ne5c
2019-12-02T22:08:18.219-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (303 > 190) bafy2bzacedtu4wcfku2wylrqtewnhwmy4axhfq547a3h7lm4nzenzpue5e3gu
2019-12-02T22:08:18.219-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (304 > 190) bafy2bzacecw2uifzv6p76l2zoy6puekf4xuwxfxvrfis4xnssh2vetkez4fcw
2019-12-02T22:08:18.220-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (305 > 190) bafy2bzacearylqr7kikba4id2uvaa4gfw5b55a4hgyehapqgvcswiwl5wifci
2019-12-02T22:08:18.221-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (306 > 190) bafy2bzacedjafgavmvncq4adhgrnkl5ysou3mknm7eedqxyopd4cb4in7mdz4
2019-12-02T22:08:18.221-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1180 > 1171) bafy2bzacebdxqvgw2nnomlp4su2omjbw7nswyu3pj5cs3qijpfoybucre36xu
2019-12-02T22:08:18.222-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1181 > 1171) bafy2bzaceaaqhrrwjgenqfdj4e4e4ldbaieqpyzc5p2yfnjl6a2uar6ouv576
2019-12-02T22:08:18.223-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1182 > 1171) bafy2bzacea2kupo5jjcmbruitqp5msoiogrmqe6n5t7fdaksl47x7hisnwfck
2019-12-02T22:08:18.224-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1183 > 1171) bafy2bzaceauivrecgqr3ykmwebasaaxjljvvn5mpq4urfp7qsbwuq4trs6nim
2019-12-02T22:08:18.224-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1184 > 1171) bafy2bzaceauy3zzusqjsqciqiizqkr33lxlraxnsxq72hbjlykfzejx5zcylq
2019-12-02T22:08:18.225-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1185 > 1171) bafy2bzacedor6fzrm7zfitdj2kc4ayykun2mqie57wojwnl6mjo7yiihiof6q
2019-12-02T22:08:18.226-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1186 > 1171) bafy2bzacebs4v3bvkyxukosdstsuhse3fifb5zze5yif7lz7mxvtwwifyohzi
2019-12-02T22:08:18.227-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (105 > 104) bafy2bzacecamlamaxcvpsu5ihuu4pkycuqaceyozfwchuwhlgcque5ivknnqy
2019-12-02T22:08:18.227-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (106 > 104) bafy2bzacec7jrkkbx2eyn4tp7mihttvdpu4gzbeb4u5fl27og4fjc5f77wbwe
2019-12-02T22:08:18.228-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1057 > 1055) bafy2bzacedygj25xqi33u64tkfdznotdnv3dyjxggosalcorali2k3pfaae7m
2019-12-02T22:08:18.229-0800	WARN	miner	miner/miner.go:387	message in mempool has too high of a nonce (1058 > 1055) bafy2bzacebdmaabcmycdwcc3ujosjdtyunqvzsbvsntbi6plyej4zxouwiukc
```

File: miner/miner.go [hunk 1]
```diff
@@ -334,14 +346,45 @@ func (m *Miner) computeTicket(ctx context.Context, addr address.Address, base *M
...
+			max = int64(m.Message.Nonce)
+			balance = types.BigSub(balance, m.Message.RequiredFunds())
+		}
 	}
 
-	msgs, err := selectMessages(context.TODO(), m.api.StateGetActor, base, pending)
+	max++ // next unapplied nonce
+
+	if max != -1 && uint64(max) != curnonce {
+		return 0, nil, xerrors.Errorf("tipset messages from %s have too low nonce %d, expected %d, h: %d", addr, max, curnonce, ts.Height())
```
Comment (magik6k) [inline miner/miner.go]: Do we want to keep all this as a sanity check?

File: miner/miner.go [hunk 2]
```diff
@@ -334,14 +346,45 @@ func (m *Miner) computeTicket(ctx context.Context, addr address.Address, base *M
...
-func (m *Miner) createBlock(base *MiningBase, addr address.Address, ticket *types.Ticket, proof *types.EPostProof) (*types.BlockMsg, error) {
+func (m *Miner) actorLookup(ctx context.Context, addr address.Address, ts *types.TipSet) (uint64, *types.BigInt, error) {
+	// TODO: strong opportunities for some caching in this method
+	act, err := m.api.StateGetActor(ctx, addr, ts)
+	if err != nil {
+		return 0, nil, xerrors.Errorf("looking up actor failed: %w", err)
+	}
 
-	pending, err := m.api.MpoolPending(context.TODO(), base.ts)
+	msgs, err := m.api.ChainGetTipSetMessages(ctx, ts.Key())
```
Comment (whyrusleeping) [inline miner/miner.go]: so i think we can remove this whole messages checking. its just a sanity check at this point, and i think we found out the real issue separately

File: miner/miner.go (hunk 1 above)
Comment (whyrusleeping) [reply miner/miner.go]: lets drop it

File: chain/gen/gen.go [hunk 3]
```diff
@@ -529,24 +540,33 @@ func IsRoundWinner(ctx context.Context, ts *types.TipSet, round int64, miner add
 		return false, nil, nil
 	}
 
-	proof, err := epp.ComputeProof(ctx, sectors, hvrf[:], winners)
+	return true, &ProofInput{
```
Comment (whyrusleeping) [inline chain/gen/gen.go]: can drop the boolean return value, and use the proofinput being non-nil for the same check

**Pull request merged**